File: 572.Subtree-of-Another-Tree/tree.c
#include <stdbool.h>

struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};

bool __is_subtree(struct TreeNode *root, struct TreeNode *sub_root) {
  if (!root && !sub_root) {
    return true;
  }

  if (!root || !sub_root) {
    return false;
  }

  if (root->val != sub_root->val) {
    return false;
  }

  // O(m)
  if (!__is_subtree(root->left, sub_root->left)) {
    return false;
  }

  // O(m)
  if (!__is_subtree(root->right, sub_root->right)) {
    return false;
  }

  return true;
}

bool isSubtree(struct TreeNode *root, struct TreeNode *subRoot) {
  if (!root) {
    return false;
  }

  // O(min(n, m))
  if (__is_subtree(root, subRoot)) {
    return true;
  }

  // O(n * m)
  if (isSubtree(root->left, subRoot)) {
    return true;
  }

  // O(n * m)
  if (isSubtree(root->right, subRoot)) {
    return true;
  }

  return false;
}

File: 235.Lowest-Common-Ancestor-of-a-Binary-Search-Tree/tree.rs
use std::{cell::RefCell, rc::Rc};

impl Solution {
    fn lca(root: Option<Rc<RefCell<TreeNode>>>, p: i32, q: i32) -> Option<Rc<RefCell<TreeNode>>> {
        let Some(node) = root else {
            return None;
        };

        let mut borrow = node.borrow_mut();
        if borrow.val == p || borrow.val == q {
            drop(borrow);
            return Some(node);
        }

        // O(n)
        let left = Self::lca(borrow.left.take(), p, q);
        let right = Self::lca(borrow.right.take(), p, q);
        drop(borrow);

        if left.is_some() && right.is_some() {
            Some(node)
        } else {
            left.or(right)
        }
    }

    pub fn lowest_common_ancestor(
        root: Option<Rc<RefCell<TreeNode>>>,
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let p = p.unwrap().borrow().val;
        let q = q.unwrap().borrow().val;

        // O(n)
        Self::lca(root, p, q)
    }
}
